src/chat.rs

```diff
@@ -2241,26 +2241,34 @@ pub async fn get_chat_msgs(
                 };
 
             Ok((
-                row.get::<_, MsgId>("id")?,
                 row.get::<_, i64>("timestamp")?,
+                row.get::<_, MsgId>("id")?,
                 !is_info_msg,
             ))
         }
     } else {
         |row: &rusqlite::Row| {
             Ok((
-                row.get::<_, MsgId>("id")?,
                 row.get::<_, i64>("timestamp")?,
+                row.get::<_, MsgId>("id")?,
                 false,
             ))
         }
     };
     let process_rows = |rows: rusqlite::MappedRows<_>| {
+        // It is faster to sort here rather than
+        // let sqlite execute an ORDER BY clause.
+        let mut sorted_rows = Vec::<(i64, MsgId, bool)>::new();
+        for row in rows {
+            let (ts, curr_id, exclude_message) = row?;
+            sorted_rows.push((ts, curr_id, exclude_message));
```

Comment:
Can someone explain why
```
sorted_rows.push(row?);
```
is consistently 2-3 percent slower than the more explicit unpacking/packing and then pushing?

Reply:
This code doesn't compile, since you can't directly collect an `Iter<sql::Result<_>>` into an `anyhow::Result<Vec<_>>` (note that the code is trying to simultaneously convert from `Collection<Result>` to `Result<Collection>` and from `sql::Result` to `anyhow::Result`).

When playing around, I wrote this code yesterday, which does the same except that it doesn't exclude excluded messages:
```rust
let collected: rusqlite::Result<Vec<(i64, MsgId, bool)>> = rows.collect();
let mut sorted_rows = collected?;
sorted_rows.sort_unstable();
```

and while looking nicer, it's 4% slower than @hpk42's code on my testing database (when I comment out the expensive `delete_expired_messages()` operation), which suggests that the imperative approach is better here.

WRT excluding messages earlier: That's also possible imperatively:
```rust
        let mut sorted_rows = Vec::new();
        for row in rows {
            let (ts, curr_id, exclude_message): (i64, MsgId, bool) = row?;
            if !exclude_message {
                sorted_rows.push((ts, curr_id));
            }
        }
        sorted_rows.sort_unstable();
```
<details>

```diff
@@ -2260,14 +2260,16 @@ pub async fn get_chat_msgs(
         let mut sorted_rows = Vec::new();
         for row in rows {
             let (ts, curr_id, exclude_message): (i64, MsgId, bool) = row?;
-            sorted_rows.push((ts, curr_id, exclude_message));
+            if !exclude_message {
+                sorted_rows.push((ts, curr_id));
+            }
         }
         sorted_rows.sort_unstable();
 
         let mut ret = Vec::new();
         let mut last_day = 0;
         let cnv_to_local = dc_gm2local_offset();
-        for (ts, curr_id, exclude_message) in sorted_rows {
+        for (ts, curr_id) in sorted_rows {
             if let Some(marker_id) = marker1before {
                 if curr_id == marker_id {
                     ret.push(ChatItem::Marker1);
@@ -2283,10 +2285,8 @@ pub async fn get_chat_msgs(
                     last_day = curr_day;
                 }
             }
-            if !exclude_message {
             ret.push(ChatItem::Message { msg_id: curr_id });
         }
-        }
         Ok(ret)
     };
```

</details>
There is no difference in performance in my benchmark, but I think it makes the code nicer to exclude them earlier.

Reply:
the collect works, I just missed a type annotation, needs to be `  .collect::<Result<Vec<_>, _>>()?;

Reply:
this is 4% faster on my machine

```diff
diff --git a/src/chat.rs b/src/chat.rs
index eb9b5212..dec20b73 100644
--- a/src/chat.rs
+++ b/src/chat.rs
@@ -2258,36 +2258,49 @@ pub async fn get_chat_msgs(
     let process_rows = |rows: rusqlite::MappedRows<_>| {
         // It is faster to sort here rather than
         // let sqlite execute an ORDER BY clause.
-        let mut sorted_rows = Vec::<(i64, MsgId, bool)>::new();
-        for row in rows {
-            let (ts, curr_id, exclude_message) = row?;
-            sorted_rows.push((ts, curr_id, exclude_message));
-        }
+        let mut sorted_rows = rows
+            .filter_map(|row| match row {
+                Ok((ts, id, exclude)) => {
+                    if exclude {
+                        None
+                    } else {
+                        Some(Ok((ts, id)))
+                    }
+                }
+                Err(err) => Some(Err(err)),
+            })
+            .collect::<std::result::Result<Vec<_>, _>>()?;
         sorted_rows.sort_unstable();
 
-        let mut ret = Vec::new();
         let mut last_day = 0;
         let cnv_to_local = dc_gm2local_offset();
-        for (ts, curr_id, exclude_message) in sorted_rows {
-            if let Some(marker_id) = marker1before {
-                if curr_id == marker_id {
-                    ret.push(ChatItem::Marker1);
+
+        let ret = sorted_rows
+            .into_iter()
+            .flat_map(|(ts, curr_id)| {
+                let mut ret = vec![];
+
+                if let Some(marker_id) = marker1before {
+                    if curr_id == marker_id {
+                        ret.push(ChatItem::Marker1)
+                    }
                 }
-            }
-            if (flags & DC_GCM_ADDDAYMARKER) != 0 {
-                let curr_local_timestamp = ts + cnv_to_local;
-                let curr_day = curr_local_timestamp / 86400;
-                if curr_day != last_day {
-                    ret.push(ChatItem::DayMarker {
-                        timestamp: curr_day,
-                    });
-                    last_day = curr_day;
+
+                if (flags & DC_GCM_ADDDAYMARKER) != 0 {
+                    let curr_local_timestamp = ts + cnv_to_local;
+                    let curr_day = curr_local_timestamp / 86400;
+                    if curr_day != last_day {
+                        ret.push(ChatItem::DayMarker {
+                            timestamp: curr_day,
+                        });
+                        last_day = curr_day;
+                    }
                 }
-            }
-            if !exclude_message {
+
                 ret.push(ChatItem::Message { msg_id: curr_id });
-            }
-        }
+                ret
+            })
+            .collect();
         Ok(ret)
     };
 ```

Reply:
This is weird. When i apply the first part (the let mut sorted_rows=) of your patch i get a 4-5% speed decrease already.

Reply:
and the second part causes another slow 4-5% down on my measurement.  It just seems that the functional iterational version is not as optimized as the plain imperative for-loop.

Reply:
or maybe it's a per-platform thing and rustc on your mac does more than on my linux?